testsuite/forge/src/backend/local/swarm.rs

```diff
@@ -0,0 +1,334 @@
+        self
+    }
+
+    pub fn dir<T: AsRef<Path>>(mut self, dir: T) -> Self {
+        self.dir = Some(dir.as_ref().into());
+        self
+    }
+
+    pub fn build(self) -> Result<LocalSwarm> {
+        let dir = if let Some(dir) = self.dir {
+            if dir.exists() {
+                fs::remove_dir_all(&dir)?;
+            }
+            fs::create_dir_all(&dir)?;
+            SwarmDirectory::Persistent(dir)
+        } else {
+            SwarmDirectory::Temporary(TempDir::new()?)
+        };
+
+        let (root_keys, validators) = ValidatorBuilder::new(&dir)
+            .num_validators(self.number_of_validators)
+            .template(self.template)
+            .build()?;
+        let diem_node_bin_path = self.diem_node_bin_path;
+
+        let validators = validators
+            .into_iter()
+            .map(|v| {
+                let node = LocalNode::new(&diem_node_bin_path, v.name, v.directory)?;
+                Ok((node.peer_id(), node))
+            })
+            .collect::<Result<HashMap<_, _>>>()?;
+
+        let validator_network_address_encryption_key = ValidatorNetworkAddressEncryptionKey {
+            key: root_keys.validator_network_address_encryption_key,
+            version: root_keys.validator_network_address_encryption_key_version,
+        };
+
+        let root_account = LocalAccount::new(
+            diem_sdk::types::account_config::diem_root_address(),
+            AccountKey::from_private_key(root_keys.root_key),
+            0,
+        );
+
+        // Designated dealer account is built to have the same private key as the TC account
+        let designated_dealer_account = LocalAccount::new(
+            diem_sdk::types::account_config::testnet_dd_account_address(),
+            AccountKey::from_private_key(
+                Ed25519PrivateKey::try_from(root_keys.treasury_compliance_key.to_bytes().as_ref())
+                    .unwrap(),
+            ),
+            0,
+        );
+
+        let treasury_compliance_account = LocalAccount::new(
+            diem_sdk::types::account_config::treasury_compliance_account_address(),
+            AccountKey::from_private_key(root_keys.treasury_compliance_key),
+            0,
+        );
+
+        Ok(LocalSwarm {
+            validators,
+            dir,
+            validator_network_address_encryption_key,
+            root_account,
+            treasury_compliance_account,
+            designated_dealer_account,
+            chain_id: ChainId::test(),
+        })
+    }
+}
+
+#[derive(Debug)]
+pub struct LocalSwarm {
+    validators: HashMap<PeerId, LocalNode>,
+    dir: SwarmDirectory,
+    validator_network_address_encryption_key: ValidatorNetworkAddressEncryptionKey,
+    // validator_swarm: DiemSwarm,
+    // vfn_swarm: DiemSwarm,
+    // pfn_swarm: DiemSwarm,
+    root_account: LocalAccount,
+    treasury_compliance_account: LocalAccount,
+    designated_dealer_account: LocalAccount,
+    chain_id: ChainId,
+    // url: String
```

Comment:
Seems incomplete, but I guess this is your building something that isn't DiemSwarm.

Is this PR a stepping stone? Because I see all the todos below here.

Reply:
Yeah this PR is quite long already so its not 100% complete yet, more to follow later :)

But i'll remove those comments